File: 프로그래머스/2/169199. 리코쳇 로봇/리코쳇 로봇.cpp
#include <string>
#include <vector>
#include <queue>
#include <iostream>

using namespace std;

int sy, sx, ey, ex, n, m;
int dir_y[] = {-1, 1, 0, 0};
int dir_x[] = {0, 0, -1, 1};

bool in_range(int y, int x, const vector<string> &board){
    if(y < 0 || y >= n || x < 0 || x >= m)
        return false;
    if(board[y][x] == 'D')
        return false;
    return true;
}

int solution(vector<string> board) {
    int answer = -1;
    n = board.size();
    m = board[0].size();
    
    vector<vector<int>> visited(board.size(), vector<int>(board[0].size(), 0));
    
    for(int i = 0; i < board.size(); i++){
        for(int j = 0; j < board[i].size(); j++){
            if(board[i][j] == 'R'){
                sy = i;
                sx = j;
            }
            
            if(board[i][j] == 'G'){
                ey = i;
                ex = j;
            }
        }
    }
    
    visited[sy][sx] = 1;
    priority_queue<pair<int, pair<int, int>>, vector<pair<int, pair<int, int>>>, greater<pair<int, pair<int, int>>>> pq;
    pq.push({0, {sy,sx}});
    
    while(!pq.empty()){
        int cnt = pq.top().first;
        int y = pq.top().second.first;
        int x = pq.top().second.second;
        pq.pop();
        
        if(y == ey && x == ex){
            answer = cnt;
            break;
        }
        
        for(int i = 0; i < 4; i++){
            int ny = y;
            int nx = x;
            int dy = dir_y[i];
            int dx = dir_x[i];
            
            while(in_range(ny+dy, nx + dx, board)){
                ny += dy;
                nx += dx;
            }
            
            if(!visited[ny][nx]){
                visited[ny][nx] = 1;
                pq.push({cnt+1, {ny, nx}});
            }
        }
    }
    
    return answer;
}

File: 프로그래머스/2/169199. 리코쳇 로봇/리코쳇 로봇.py
from collections import deque
from queue import PriorityQueue

def in_range(y, x, n, m, board):
    if 0 <= y < n and 0 <= x < m and board[y][x] != 'D':
        return True
    return False

def solution(board):
    answer = -1
    n = len(board)
    m = len(board[0])
    
    s_pos = None
    e_pos = None
    
    for i in range(n):
        for j in range(m):
            if board[i][j] == 'R':
                s_pos = (i, j)
                break
    
    q = PriorityQueue()
    q.put((0, [s_pos[0], s_pos[1]]))
    visited = [[-1] * m for _ in range(n)]
    visited[s_pos[0]][s_pos[1]] = 0
    dx = [0, 0, -1, 1]
    dy = [-1, 1, 0, 0]
    
    while not q.empty():
        cnt, (y, x) = q.get()
        
        if board[y][x] == 'G':
            answer = cnt
            break
        
        for i in range(4):
            ny = y
            nx = x
            
            while in_range(ny + dy[i], nx + dx[i], n, m, board):
                ny += dy[i]
                nx += dx[i]

            if visited[ny][nx] == -1:
                visited[ny][nx] = cnt + 1
                q.put((cnt + 1, (ny, nx)))
                
    return answer